### ISSUE TITLE
A special case of

### ISSUE THREAD

**Issue by tooyoungtoosimpo** [eclipse-cyclonedds/cyclonedds#1701]

**Comment by tooyoungtoosimpo** [comment]
_In the case where the reliability QoS kind is RELIABLE then in steady-state, defined as the situation where the DataWriter
does not write new samples for a period “long” compared to the minimum_separation, the system should guarantee delivery
the last sample to the DataReader._

This case is unimplemented.

**Comment by eboasson** [comment]
That's correct, the documentation should be improved to reflect that 😀

If you don't reduce the sample rate at the writer side, then time-based filter is really equivalent to a keep-last reader and polling. So the cases where it might come in useful are those where the writer refrains from sending data.

The middleware by definition doesn't know when the next sample will come. Is it ok to just delay samples? What if the writer publishes every T-𝜀 where the writer has a time-based filter setting of T? Should the writer then send it always? Then at what value of 𝜀 do things change? How much additionally latency in such a situation would be acceptable?

This is of course just an instance of:

> [...] does not write new samples for a period “long” compared [...]

which is a really a case of a standards body writing something without thinking it through[^1]. That's not a definition. How long is "long"? 2x, 10x, 100x? What if the reader sets a minimum separation of 1ns? Is it then suddenly 100ms? Can you risk packet drops on the network, or the writer getting deleted? Is the writer really to track all this separately for a bunch of different readers? How does it interact with multicast? What if I have 4 matched readers with minimum separation values of 1ns, 1µs, 1ms, 1s? And what if the readers also use content filters? (And does it even apply to historical data from a transient-local writer?)

There is a possible use case for data rate reduction outside LANs, for example when bridging networks. One could think this feature would help with that, but in practice it works must better to just build a [gateway](https://zenoh.io/blog/2021-09-28-iac-experiences-from-the-trenches/) that does the rate reduction internally.

In short I consider the time-based filter a horribly underspec'd misfeature that only adds to the bloat of a DDS implementation. I would have preferred leaving the entire QoS setting out; what we have was the compromise. But I am always interested in arguments that prove me wrong!

[^1]: Or pushing in a feature from a commercial pub/sub messaging middleware from the mid '90s that presumably never was defined properly either.
